datashader/datatypes.py

```diff
@@ -0,0 +1,364 @@
+            Indices to be taken.
+        allow_fill : bool, default False
+            How to handle negative values in `indices`.
+
+            * False: negative values in `indices` indicate positional indices
+              from the right (the default). This is similar to
+              :func:`numpy.take`.
+
+            * True: negative values in `indices` indicate
+              missing values. These values are set to `fill_value`. Any other
+              other negative values raise a ``ValueError``.
+
+        fill_value : any, default None
+            Fill value to use for NA-indices when `allow_fill` is True.
+
+        Returns
+        -------
+        RaggedArray
+
+        Raises
+        ------
+        IndexError
+            When the indices are out of bounds for the array.
+        """
+        if allow_fill:
+            sequence = [self[i] if i >= 0 else fill_value
+                        for i in indices]
+        else:
+            sequence = [self[i] for i in indices]
+
+        return RaggedArray(sequence, dtype=self.flat_array.dtype)
+
+    def copy(self, deep=False):
+        """
+        Return a copy of the array.
+
+        Parameters
+        ----------
+        deep : bool, default False
+            Also copy the underlying data backing this array.
+
+        Returns
+        -------
+        RaggedArray
+        """
+        data = dict(
+            mask=self.mask,
+            flat_array=self.flat_array,
+            start_indices=self.start_indices)
+
+        if deep:
+            # Copy underlying numpy arrays
+            data = {k: v.copy() for k, v in data.items()}
+
+        return RaggedArray(data)
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        """
+        Concatenate multiple RaggedArray instances
+
+        Parameters
+        ----------
+        to_concat : list of RaggedArray
+
+        Returns
+        -------
+        RaggedArray
+        """
+        # concat masks
+        mask = np.hstack(ra.mask for ra in to_concat)
+
+        # concat flat_arrays
+        flat_array = np.hstack(ra.flat_array for ra in to_concat)
+
+        # offset and concat start_indices
+        offsets = np.hstack([
+            [0],
+            np.cumsum([len(ra.flat_array) for ra in to_concat[:-1]])])
+
+        start_indices = np.hstack([ra.start_indices + offset
+                                   for offset, ra in zip(offsets, to_concat)])
+
+        return RaggedArray(dict(
+            mask=mask, flat_array=flat_array, start_indices=start_indices))
+
+    @property
+    def dtype(self):
+        return RaggedDtype()
+
+    @property
+    def nbytes(self):
+        """
+        The number of bytes needed to store this object in memory.
+        """
+        return (self._flat_array.nbytes +
+                self._start_indices.nbytes +
+                self._mask.nbytes)
+
+    def astype(self, dtype, copy=True):
+        if isinstance(dtype, RaggedDtype):
```

Comment:
Note: some people may hope that `.astype('category')` (or any other extension dtype) works here.

I'm not sure whether it makes sense to support that. But just thought I'd make you aware that `ExtensionArray.astype` *may* want to support ExtensionArray[T] -> ExtensionArray[S] conversions.

Reply:
Interesting. Yeah this should at least support `RaggedArray[T]` -> `RaggedArray[S]` to support casting the datatype of the underlying `flat_array`.

If it's some other extension type, would it be reasonable to default to passing `np.array(self)` to the constructor of the other extension array type?

Reply:
> If it's some other extension type, would it be reasonable to default to passing np.array(self) to the constructor of the other extension array type?

Pandas doesn't make any assumptions about any subclasses `ExtensionArray.__init__`. If you want to, something like

```python
dtype = pandas_dtype(dtype)

if is_extension_array_dtype(dtype):
    dtype.construct_array_type()._from_sequence(np.asarray(self))
```

should work for any ExtensionArray that can handle your scalar type (which may not be many, but best to let them error if they can't handle it.

FYI, there's discussion on this here: https://github.com/pandas-dev/pandas/issues/22384 If you have any thoughts, it'd be good to hear them.

Reply:
Is there a non-private import for `is_extension_array_dtype`?  I'm only finding it through `pandas.core.dtypes.common`

Reply:
Sorry. https://github.com/pandas-dev/pandas/pull/24771, so `pandas.core` will have to do for the RC.